### PR TITLE
fix: [M3-7262] - Create Firewall Drawer opens in the same tab in the Linode Create Flow

### PR DESCRIPTION
**Pull request by tyler-akamai**
## Description 📝
Currently there is a 'Create Firewall' button in the Firewall Panel that when clicked, opens the Create Firewall Drawer in another tab. This should open the Create Firewall drawer in the same tab, in the same page.

## Changes  🔄
List any change relevant to the reviewer.
- Firewall Panel's 'Create Firewall' button opens the Firewall Create Drawer in the same tab
- Once the Firewall is selected, it becomes the value of the dropdown
- Dropdown header in Create Firewall Drawer 'Linodes' --> 'Additional Linodes (Optional)'

| Before  | After   |
| ------- | ------- |
| <video src="https://github.com/linode/manager/assets/139489745/42ca1ea3-5907-4429-ab8e-a1746b9608e2" /> | <video src="https://github.com/linode/manager/assets/139489745/d2361b15-c521-4a34-b743-284b125c9e58" /> |

### Prerequisites
- No required prerequisites

### Reproduction steps
- Navigate to http://localhost:3000/linodes/create
- Scroll down to the Firewall Panel
- Click on the 'Create VPC' link
- Ensure that it brings you to a new tab

### Verification steps 
Navigate to http://localhost:3000/linodes/create
Ensure that:
- [ ] The default Firewall is still 'None'
- [ ] When you choose a Firewall, the summary section at the bottom gets updated ('Firewall Assigned' gets added)
- [ ] When clicking on 'None', the 'Firewall Assigned' gets removed from the summary section
- [ ] When clicking on the 'Create Firewall' link, the Create Firewall Drawer appears in the same tab
- [ ] The Create Firewall Drawer has the text 'Additional Linodes (Optional)', while the Create Firewall Drawer from the Firewall Landing page has the text 'Linodes'
- [ ] Add an additional Linode and ensure it is added when the Firewall is created
- [ ] After creating a new Firewall, that Firewall is selected in the dropdown (in the firewall panel) and the summary section got updated
- [ ] After creating the Linode, the Linode is properly assigned to the Firewall in the Firewall Landing page. **Test this in `development`, it is blocked in `prod` by an api feature flag**

## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [x] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [x] 🧪 Providing/Improving test coverage
- [x] 🔐 Removing all sensitive information from the code and PR description
- [x] 🚩 Using a feature flag to protect the release
- [x] 👣 Providing comprehensive reproduction steps
- [x] 📑 Providing or updating our documentation
- [x] 🕛 Scheduling a pair reviewing session
- [x] 📱 Providing mobile support
- [x] ♿  Providing accessibility support

### CODE REVIEW

File: packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx [hunk 1]
```diff
@@ -17,6 +17,17 @@ interface Props {
...
   const { handleFirewallChange, helperText } = props;
 
+  const [isDrawerOpen, setIsDrawerOpen] = React.useState(false);
+
+  const handleCreateFirewallClick = () => {
+    setIsDrawerOpen(true);
+  };
+
+  const handleFirewallCreated = (id: number, label: string) => {
+    setDropdownValue({ label, value: id });
```
Comment (tyler-akamai) [inline packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx:27]: When the user creates a new Firewall, this code is responsible for setting the value of the dropdown to the recently created Firewall.

File: packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx [hunk 2]
```diff
@@ -44,19 +60,34 @@ export const SelectFirewallPanel = (props: Props) => {
...
+          onClick={(e) => {
+            e.preventDefault();
+            handleCreateFirewallClick();
+          }}
+          to="#"
+        >
           Create Firewall
         </StyledCreateLink>
+        <CreateFirewallDrawer
+          label="Additional Linodes (Optional)"
```
Comment (tyler-akamai) [inline packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx:86]: UX wanted to ensure the user:
- Wasnt looking for current Linode in the dropdown --> 'Additional'
- Knew that this input was optional --> '(Optional)'

File: packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx [hunk 3]
```diff
@@ -14,9 +14,22 @@ interface Props {
   helperText: JSX.Element;
 }
 
+export const createFirewallLabel = 'Additional Linodes (Optional)';
```
Comment (tyler-akamai) [inline packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx:17]: UX wanted to ensure the user:
- Wasnt looking for current Linode in the dropdown --> 'Additional'
- Knew that the input was optional --> '(Optional)'

**Review by mjac0bs — COMMENTED** [review]
Thank you for this awesome testing checklist!

I verified:
- [x] The default Firewall is still 'None'
- [x]  When you choose a Firewall, the summary section at the bottom gets updated ('Firewall Assigned' gets added)
- [x]  When clicking on 'None', the 'Firewall Assigned' gets removed from the summary section
- [x]  When clicking on the 'Create Firewall' link, the Create Firewall Drawer appears in the same tab
- [X]  The Create Firewall Drawer has the text 'Additional Linodes (Optional)', while the Create Firewall Drawer from the Firewall Landing page has the text 'Linodes'
- [x]  Add an additional Linode and ensure it is added when the Firewall is created
- [x]  After creating a new Firewall, that Firewall is selected in the dropdown (in the firewall panel) and the summary section got updated
- [ ]  After creating the Linode, the Linode is properly assigned to the Firewall in the Firewall Landing page.

For the last item, I'm not seeing the linode in the create flow be assigned to the firewall -- just the "additional linode" is assigned. Based on your screencap of this branch and my understanding of the acceptance criteria, I would have expected to see both `longview-linode-do-not-delete` and `debian-us-sea` listed as assigned to firewall `testing123`, but `debian-us-sea` was not. Have you observed any issues with `handleFirewallChange`?

https://github.com/linode/manager/assets/114685994/bf81fe52-ea62-471d-a712-1d760ca9d99b

**Comment by tyler-akamai** [comment]
> Thank you for this awesome testing checklist!
> 
> I verified:
> 
> * [x]  The default Firewall is still 'None'
> * [x]   When you choose a Firewall, the summary section at the bottom gets updated ('Firewall Assigned' gets added)
> * [x]   When clicking on 'None', the 'Firewall Assigned' gets removed from the summary section
> * [x]   When clicking on the 'Create Firewall' link, the Create Firewall Drawer appears in the same tab
> * [x]   The Create Firewall Drawer has the text 'Additional Linodes (Optional)', while the Create Firewall Drawer from the Firewall Landing page has the text 'Linodes'
> * [x]   Add an additional Linode and ensure it is added when the Firewall is created
> * [x]   After creating a new Firewall, that Firewall is selected in the dropdown (in the firewall panel) and the summary section got updated
> * [ ]   After creating the Linode, the Linode is properly assigned to the Firewall in the Firewall Landing page.
> 
> For the last item, I'm not seeing the linode in the create flow be assigned to the firewall -- just the "additional linode" is assigned. Based on your screencap of this branch and my understanding of the acceptance criteria, I would have expected to see both `longview-linode-do-not-delete` and `debian-us-sea` listed as assigned to firewall `testing123`, but `debian-us-sea` was not. Have you observed any issues with `handleFirewallChange`?
> 
>  Screen.Recording.2023-10-18.at.10.07.57.AM.mov

Good catch! I had to ask @dwiley-akamai on this one. From my understanding this functionality is blocked by an API feature flag. You can test this in `develop` since the feature flag is turned on, but it is not turned on in prod yet.

File: packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx [hunk 4]
```diff
@@ -30,6 +43,11 @@ export const SelectFirewallPanel = (props: Props) => {
     value: -1,
   });
 
+  const [dropdownValue, setDropdownValue] = React.useState<{
+    label: string;
+    value: number;
+  }>(firewallsDropdownOptions[0]);
```
Comment (bnussman-akamai) [inline packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx:49]: I feel like we shouldn't have to do:
```ts
  firewallsDropdownOptions.unshift({
    label: 'None',
    value: -1,
  });
```
It should be okay for the dropdown to default to a `value` of `null`.

File: packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx [hunk 5]
```diff
@@ -44,19 +62,34 @@ export const SelectFirewallPanel = (props: Props) => {
       <Stack>
         {helperText}
         <Select
```
Comment (bnussman-akamai) [inline packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx:64]: I think we should go ahead and switch this to a `Autocomplete`, it should be close to a drop-in replacement.

**Comment by jaalah-akamai** [comment]
I verified the checklist above and the last item as well. This works in dev, not locally though.

Screenshot 2023-10-20 at 9 11 03 AM

**Review by jaalah-akamai — APPROVED** [review]
Looks good pending changes requested from @bnussman-akamai

File: packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx [hunk 6]
```diff
@@ -25,10 +39,15 @@ export const SelectFirewallPanel = (props: Props) => {
...
-  });
+  // firewallsDropdownOptions.unshift({
+  //   label: 'None',
+  //   value: -1,
+  // });
+
+  const [dropdownValue, setDropdownValue] = React.useState<{
+    label: string;
+    value: number;
+  } | null>(null);
```
Comment (bnussman-akamai) [inline packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx:50]: I just realized, we shouldn't need state here. We should have a prop on this component called `selectedFirewallId: number` which we should be using in place of the state

```suggestion
```

**Review by bnussman-akamai — CHANGES_REQUESTED** [review]

**Review by mjac0bs — APPROVED** [review]
- [x]  After creating the Linode, the Linode is properly assigned to the Firewall in the Firewall Landing page.

Verified in `dev` environment, thanks for the clarification!

Functionality looks good. You caught the commented code I was about to note. Other than Banks' feedback and my earlier changelog suggestion, this LGTM.

File: packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx [hunk 7]
```diff
@@ -43,20 +59,30 @@ export const SelectFirewallPanel = (props: Props) => {
...
+            handleFirewallChange(selection?.value ?? -1);
+          }}
           errorText={error?.[0].reason}
-          isClearable={false}
-          isLoading={isLoading}
           label="Assign Firewall"
-          noOptionsMessage={() => 'Create a Firewall to assign to this Linode.'}
-          onChange={(selection) => handleFirewallChange(selection.value)}
+          loading={isLoading}
+          noOptionsText="Create a Firewall to assign to this Linode."
```
Comment (bnussman-akamai) [inline packages/manager/src/components/SelectFirewallPanel/SelectFirewallPanel.tsx:69]: Non-bocking feedback: 

This copy feels a bit weird to me. It makes it should like there should be a call to action but there isn't one in the select component.

File: packages/manager/src/features/Linodes/LinodesCreate/LinodeCreate.tsx [hunk 8]
```diff
@@ -381,7 +381,7 @@ export class LinodeCreate extends React.PureComponent<
       });
     }
 
-    if (this.props.firewallId !== undefined && this.props.firewallId !== -1) {
+    if (this.props.firewallId != null && this.props.firewallId !== -1) {
```
Comment (bnussman-akamai) [inline packages/manager/src/features/Linodes/LinodesCreate/LinodeCreate.tsx:384]: we probably want to use `!==` and maybe check for undefined too?

**Review by bnussman-akamai — APPROVED** [review]
Thanks for addressing the feedback! Great change!

**Pull request merged**